libs/langchain/langchain/llms/fireworks.py

```diff
@@ -66,43 +70,89 @@ def _llm_type(self) -> str:
         """Return type of llm."""
         return "fireworks"
 
-    def _call(
+    def _generate(
         self,
-        prompt: str,
+        prompts: List[str],
         stop: Optional[List[str]] = None,
         run_manager: Optional[CallbackManagerForLLMRun] = None,
         **kwargs: Any,
-    ) -> str:
-        """Run the LLM on the given prompt and input."""
-        params: dict = {
+    ) -> LLMResult:
+        """Call out to Fireworks endpoint with k unique prompts.
+        Args:
+            prompts: The prompts to pass into the model.
+            stop: Optional list of stop words to use when generating.
+        Returns:
+            The full LLM output.
+        """
+        params = {
             "model": self.model,
-            "prompt": prompt,
             **self.model_kwargs,
         }
-        response = completion_with_retry(
-            self, run_manager=run_manager, stop=stop, **params
-        )
+        sub_prompts = self.get_batch_prompts(params, prompts, stop)
+        choices = []
+        for _prompts in sub_prompts:
+            response = completion_with_retry_batching(
+                self, self.use_retry, prompt=_prompts, run_manager=run_manager, **params
+            )
+            choices.extend(response)
 
-        return response.choices[0].text
+        return self.create_llm_result(choices, prompts)
 
-    async def _acall(
+    async def _agenerate(
         self,
-        prompt: str,
+        prompts: List[str],
         stop: Optional[List[str]] = None,
         run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
         **kwargs: Any,
-    ) -> str:
-        """Run the LLM on the given prompt and input."""
+    ) -> LLMResult:
+        """Call out to Fireworks endpoint async with k unique prompts."""
         params = {
             "model": self.model,
-            "prompt": prompt,
             **self.model_kwargs,
         }
-        response = await acompletion_with_retry(
-            self, run_manager=run_manager, stop=stop, **params
-        )
+        sub_prompts = self.get_batch_prompts(params, prompts, stop)
+        choices = []
+        for _prompts in sub_prompts:
+            response = await acompletion_with_retry_batching(
+                self, self.use_retry, prompt=_prompts, run_manager=run_manager, **params
+            )
+            choices.extend(response)
+
+        return self.create_llm_result(choices, prompts)
 
-        return response.choices[0].text
+    def get_batch_prompts(
+        self,
+        params: Dict[str, Any],
+        prompts: List[str],
+        stop: Optional[List[str]] = None,
+    ) -> List[List[str]]:
+        """Get the sub prompts for llm call."""
+        if stop is not None:
+            if "stop" in params:
+                raise ValueError("`stop` found in both the input and default params.")
+            params["stop"] = stop
```

Comment:
@tjaffri I put stop to params here. Does it make sense?

Reply:
It probably works, but you are now doing things differently from how it is done elsewhere for fireworks... which will be confusing to other maintainers and might lead to bugs later (remember we have had bugs with stops words before with fireworks). For example, look at:

1. Lines 141, 160, 203 and 232 in the fireworks chat model... it just passes the stop words as a named param to completion_with_retry
2. Lines 171 and 192 in the current file we are discussing (the fireworks non-chat model)... those also pas the stop words as named parameters to completion_with_retry

So I think for batching, just stick to the same... rather than doing this in get_batch_prompts just pass stop into completion_with_retry for batch and abatch... problem solved.

Reply:
sounds good, will do this to avoid confusion.